### ISSUE TITLE
Collection constructor poses issue when trying to encapsulate an array

### ISSUE THREAD

**Issue by krtek4** [widmogrod/php-monads#31]
The current constructor of the `Collection` monad poses some issues when trying to encapsulate an array. If I want to store an empty array inside the collection, I have to do `Collection::of([[]])` which is not isomorphic with say `Maybe` where I just do `Just::of([])`.

This is problematic for my implementation of `filterM` (see #26) where we try to construct the containing monad dynamically but we need a "special case" for Collection as it don't work right now.

One option could be to change the constructor of `Collection` to use `func_get_args` and never assume that a traversable passer as parameter should be stored as is.

What do you think ?

PS: I hope my description of the issue is clear.

**Comment by krtek4** [comment]
I propose the following if it is ok for you :

``` php

class Listt {
    public function __construct() {
        $this->value = func_get_args();
    }

    public static function of()
    {
        $list = new static();
        $list->value = func_get_args();
    }
}
```

And you create an instance like this :

``` php

$list = new Listt(); // will contain an empty array

$list = new Listt([]); // will contain an empty array inside an array

$list = new List(1, 2, 3, 4); // will contain [1, 2, 3, 4]
```

This makes more sense in my opinion than passing an array to the constructor and it is more flexible.

Also, I am not 100% sure, but I think the `traverse` implementation you introduced in #43 is broken if the `callable` returns a instance of `Listt`. Maybe adding a test for it would be a good idea. I could also add a test for `filterM` to demonstrate the issue.

What do you think ?

**Comment by syhol** [comment]
I like it but i think it would look nicer if it used variadics instead of func_get_args. Does this project support 5.4?

**Comment by syhol** [comment]
...i mean 5.6 :)

**Comment by krtek4** [comment]
If the lower supported PHP version is 5.6, variadic would be nice yes. It would allow to write `of` more cleanly.

@widmogrod what do you think ? What is the PHP version you want to support ?

**Comment by widmogrod** [comment]
@syhol @krtek4 currently minimum version of PHP is 5.6 - https://github.com/widmogrod/php-functional/blob/master/.travis.yml#L3 - although, not for long http://php.net/supported-versions.php

**Comment by widmogrod** [comment]
@krtek4 you highlighted interesting issue. 
I described my vision of `Listt` & problem with traversable types in php in this issue https://github.com/widmogrod/php-functional/issues/46 

Proposed solution by you

``` php
$list = new List(1, 2, 3, 4);
```

rise few questions:
- `[a]` how to adapt `Traversable` types like `[]`, `Iterator`, `Generator`,`IteratorAggregate`,... ?
- `[b]` how many times in real word application list is explicitly declared like so: `new List(1, 2, 3, 4);`?

**Comment by widmogrod** [comment]
I believe it was solved by this issue  #46
